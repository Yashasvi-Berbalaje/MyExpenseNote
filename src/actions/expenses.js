import {v4 as uuidv4} from 'uuid';

const AddExpense = ({
    description='',
    note='',
    amount=0,
    createdAt=0
}={}) =>({
    type:'ADD',
    expenses:{
       id:uuidv4(),
       description,
       note,
       amount,
       createdAt
    }
})

const removeExpense = ({id}={}) =>({
    type:'REMOVE',
    id
})

const EditExpense = (id,update) =>({
    type:'EDIT_EXPENSE',
    id,
    update
})

export {AddExpense,removeExpense,EditExpense};