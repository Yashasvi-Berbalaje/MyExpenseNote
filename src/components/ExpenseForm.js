import React, { Component } from 'react';
import 'react-dates/initialize';
import moment from 'moment';
import {SingleDatePicker} from  'react-dates';

export default class ExpenseForm extends Component{
    constructor(props){
        super(props);
        this.state={
            description:props.expense? props.expense.description:'',
            expenseNote:props.expense? props.expense.note:'',
            amount:props.expense?(props.expense.amount/100).toString():'',
            createdAt:props.expense?moment(props.expense.createdAt):moment(),
            calenderFocus:false,
            error:''
        }

    }
  

    onDescriptionChange = (event) =>{
        const description=event.target.value
        this.setState(()=>({description}))
    }
    onAddExpenseNoteChange = (event) =>{
        const expenseNote=event.target.value
        this.setState(()=>({expenseNote}))
    }
    onAmountChange = (event) =>{
        const amount=event.target.value
        if(!amount||amount.match(/^\d{1,}(\.\d{0,2})?$/)){
        this.setState(()=>({amount}))
    }
    }
    onDateChange = (createdAt) =>{
        if(createdAt)
        {
            this.setState(()=>({createdAt}));
        }
        
    }
    onFocusChange = ({focused}) =>{
        this.setState(()=>({calenderFocus:focused}))
    }
    onSubmit = (event) =>{
        event.preventDefault();
        if(!this.state.description||!this.state.amount){
            this.setState(()=>{
                return{
                    error:'please enter all required fields'
                }
            })
        }
        else{
            this.setState(()=>{
                return{
                    error:''
                }
            })
         this.props.onSubmit({
             description:this.state.description,
             amount:parseFloat(this.state.amount,10)*100,
             createdAt:this.state.createdAt.valueOf(),
             note:this.state.expenseNote
         })
        }

    }

    render(){
        return(
            <div>
                <p>{this.state.error}</p>
                <form onSubmit={this.onSubmit}>
               <input type='text' placeholder='Description' autoFocus value={this.state.description} onChange={this.onDescriptionChange} ></input>
               <input type='number' placeholder='Amount' value={this.state.amount} onChange={this.onAmountChange} ></input>
               <SingleDatePicker
                 date={this.state.createdAt}
                 onDateChange={this.onDateChange}
                 focused={this.state.calenderFocus}
                 onFocusChange={this.onFocusChange}
                 numberOfMonths={1}
                 isOutsideRange={()=>false}       
               />
               <textarea placeholder='Add a note on expense' value={this.expenseNote} onChange={this.onAddExpenseNoteChange}></textarea>
               <button>Submit</button> 
               </form>
            </div>
        )
    }
}