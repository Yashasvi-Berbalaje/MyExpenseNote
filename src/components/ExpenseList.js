import React from 'react';
import {connect} from  'react-redux';
import ExpenseListItem from './ExpenseListItem';
import  SelectorExpense  from '../selectors/expenseSelector';


export const ExpenseList = (props) =>(
    <div>
    {
        props.expenses.length ===0 ?
      (<p>No expenses</p>):(props.expenses.map((expense)=>(<ExpenseListItem key={expense.id} {...expense}/>)))  
    }
    </div>
    )


const mapStateToProps=(state)=>{
    return{
        expenses:SelectorExpense(state.expenses,state.filter)
   };
}

export default connect(mapStateToProps)(ExpenseList)