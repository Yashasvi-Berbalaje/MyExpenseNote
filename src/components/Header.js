import {NavLink} from 'react-router-dom';
import React from 'react';
 
  const Header = () =>(
    <div>
    <h1>Expensify</h1>
    <NavLink to='/Dashboard' activeClassName='is-active'>Home</NavLink>
    <NavLink to='/expense' activeClassName='is-active'>AddExpense</NavLink>
    </div>
  );
  export default Header;