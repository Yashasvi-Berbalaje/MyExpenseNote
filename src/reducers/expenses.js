
const expenseReducerDefaultState=[]
const expenseReducer = (state=expenseReducerDefaultState,action) => {
     switch(action.type){
        case 'ADD':
             return[...state,action.expenses];
        case 'REMOVE':
        return (state.filter(({id}) => id!==action.id));
        case 'EDIT_EXPENSE':
        return(
         state.map((expense)=>{
            if(expense.id===action.id)
            {
                
               return{...expense,...action.update} 
            }
            else{
                return expense;
            }
         })   
        )
         default:
             return state;
     }
}

export default expenseReducer;