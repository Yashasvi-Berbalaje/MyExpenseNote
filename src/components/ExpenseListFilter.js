import React, { Component } from 'react';
import {connect} from 'react-redux';
import {setTextFilter,sortByDate,sortByAmount,setStartDate,setEndDate} from '../actions/filters';
import {DateRangePicker} from 'react-dates';
import {v4 as uuidv4} from 'uuid';


export class ExpenseListFilter extends Component{

    state = {
        calenderFocus:null
    }
    onDatesChange=({startDate,endDate})=>{
        this.setState(()=>{
            this.props.setStartDate(startDate)
            this.props.setEndDate(endDate)
        })
    }
    onFocusChange=(focusedInput)=>{
      this.setState({calenderFocus:focusedInput})
    }
    onTextChange=(event)=>{
        this.props.setTextFilter(event.target.value)
    }
    onSortChange=(event)=>{
        event.target.value==='date'? this.props.sortByDate():this.props.sortByAmount();
   }
    render(){
        return(
            <div>
            <input type='text' value={this.props.filters.text}
            onChange={this.onTextChange}
            />
            <select value={this.props.filters.sortBy} onChange={this.onSortChange}>
                <option value='date'>Date</option>
                <option value='amount'>Amount</option>
            </select>
            <DateRangePicker
            startDate={this.props.filters.startDate}
            startDateId={uuidv4()}
            endDateId={uuidv4()}
            endDate={this.props.filters.endDate}
            onDatesChange={this.onDatesChange}
            focusedInput={this.state.calenderFocus}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={()=>false}
            />
            </div>
        )
    }
}


const mapStateToProps = (state) =>
    ({
        filters:state.filter
    })

const mapDispatchToProps = (dispatch) =>({
      setTextFilter:(text)=>dispatch(setTextFilter(text)),
      sortByDate:()=>dispatch(sortByDate()),
      sortByAmount:()=>dispatch(sortByAmount()),
      setStartDate:(startDate)=>dispatch(setStartDate(startDate)),
      setEndDate:(endDate)=>dispatch(setEndDate(endDate))
})



export default connect(mapStateToProps,mapDispatchToProps)(ExpenseListFilter)