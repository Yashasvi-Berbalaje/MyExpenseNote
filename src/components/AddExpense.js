import React, { Component } from 'react';
import {connect} from 'react-redux';
import {AddExpense} from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

export class AddExpenseComponent extends Component{
  onSubmit=(expense)=>{
    this.props.AddExpense(expense)
    this.props.history.push('/Dashboard');
  }
  render(){
    return(
    <div>
     <h1>Add Expense</h1>
     <ExpenseForm onSubmit={this.onSubmit}/>
    </div>
    )
  }
}


  const mapDispatchToProps=(dispatch)=>({
    AddExpense:(expense)=>dispatch(AddExpense(expense))
  })

export default connect(undefined,mapDispatchToProps)(AddExpenseComponent);