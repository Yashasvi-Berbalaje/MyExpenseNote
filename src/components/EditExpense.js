import React, { Component } from 'react';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {EditExpense, removeExpense} from '../actions/expenses';

export class EditExpenseComponent extends Component{
  onSubmit=(expense)=>{
    this.props.editExpense(this.props.expense.id,expense);
    this.props.history.push('/Dashboard');
  }
  onRemove=()=>{
    this.props.removeExpense({id:this.props.expense.id});
    this.props.history.push('/Dashboard');
}

  render(){
    return(
    <div>
     <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit}/> 
      <button onClick={this.onRemove}>Remove</button>
    </div>
    )
  }
}


  const mapStateToProps = (state,props) =>{
     return{
       expense:state.expenses.find((expense)=>expense.id===props.match.params.id)
  }
  }
  const mapDispatchToProps = (dispatch) =>{
      return{
        editExpense:(id,expense)=>dispatch(EditExpense(id,expense)),
        removeExpense:(id)=>dispatch(removeExpense(id))
      }
  }

 

export default connect(mapStateToProps,mapDispatchToProps)(EditExpenseComponent);