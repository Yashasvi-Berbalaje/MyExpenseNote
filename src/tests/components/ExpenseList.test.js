import React from 'react';
import {shallow } from 'enzyme';
import {ExpenseList} from '../../components/ExpenseList';
import expenses from '../fixtures/expenses';
import toJSON from 'enzyme-to-json';

test('Test case for Expense List with Expenses',()=>{
    const action=shallow(<ExpenseList expenses={expenses}/>)
    expect(toJSON(action)).toMatchSnapshot()
})
test('Test case for Expense List item without Expenses',()=>{
    const action=shallow(<ExpenseList expenses={[]}/>)
    expect(toJSON(action)).toMatchSnapshot()
})