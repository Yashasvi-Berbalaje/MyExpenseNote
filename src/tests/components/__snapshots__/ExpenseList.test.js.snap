// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test case for Expense List item without Expenses 1`] = `
<div>
  <p>
    No expenses
  </p>
</div>
`;

exports[`Test case for Expense List with Expenses 1`] = `
<div>
  <ExpenseListItem
    amount={125}
    createdAt={0}
    description="Gum"
    id="1"
    note=""
  />
  <ExpenseListItem
    amount={10000}
    createdAt={-345600000}
    description="Rent"
    id="2"
    note=""
  />
  <ExpenseListItem
    amount={3000}
    createdAt={345600000}
    description="credit card"
    id="3"
    note=""
  />
</div>
`;
